tmail-backend/webadmin/webadmin-mailbox/src/main/java/com/linagora/tmail/webadmin/cleanup/CleanupTrashService.java

```diff
@@ -0,0 +1,114 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(CleanupTrashService.class);
+    private static final Map<String, Period> mapPeriodSettingToDuration = ImmutableMap.of(JmapSettings.weeklyPeriod(), Period.ofWeeks(1),
+        JmapSettings.monthlyPeriod(), Period.ofMonths(1));
+
+    private final UsersRepository usersRepository;
+    private final JmapSettingsRepository jmapSettingsRepository;
+    private final StoreMailboxManager storeMailboxManager;
+    private final SystemMailboxesProvider systemMailboxesProvider;
+    private final Clock clock;
+
+    @Inject
+    public CleanupTrashService(UsersRepository usersRepository, JmapSettingsRepository jmapSettingsRepository, StoreMailboxManager storeMailboxManager,
+                               SystemMailboxesProvider systemMailboxesProvider, Clock clock) {
+        this.usersRepository = usersRepository;
+        this.jmapSettingsRepository = jmapSettingsRepository;
+        this.storeMailboxManager = storeMailboxManager;
+        this.systemMailboxesProvider = systemMailboxesProvider;
+        this.clock = clock;
+    }
+
+    public Mono<Task.Result> cleanupTrash(RunningOptions runningOptions) {
+        return Flux.from(usersRepository.listReactive())
+            .transform(ReactorUtils.<Username, Task.Result>throttle()
+                .elements(runningOptions.getUsersPerSecond())
+                .per(Duration.ofSeconds(1))
+                .forOperation(this::cleanupTrashForSingleUser))
+            .reduce(Task.Result.COMPLETED, Task::combine)
+            .onErrorResume(e -> {
+                LOGGER.error("Error while accessing users from repository", e);
+                return Mono.just(Task.Result.PARTIAL);
+            });
+    }
+
+    private Mono<Task.Result> cleanupTrashForSingleUser(Username username) {
+        return Mono.from(jmapSettingsRepository.get(username))
+            .flatMap(jmapSettings -> cleanupTrashForSingleUser(username, jmapSettings))
+            .defaultIfEmpty(Task.Result.COMPLETED)
+            .onErrorResume(e -> {
+                LOGGER.error("Error while cleaning trash for user {}", username, e);
+                return Mono.just(Task.Result.PARTIAL);
+            });
+    }
+
+    private Mono<Task.Result> cleanupTrashForSingleUser(Username username, JmapSettings jmapSettings) {
+        if (jmapSettings.trashCleanupEnabled()) {
+            return cleanupTrashForSingleUserWithSpecificPeriod(username,
+                mapPeriodSettingToDuration.getOrDefault(jmapSettings.trashCleanupSetting(), Period.ofMonths(1)));
+        } else {
+            return Mono.just(Task.Result.COMPLETED);
+        }
+    }
+
+    private Mono<Task.Result> cleanupTrashForSingleUserWithSpecificPeriod(Username username, Period period) {
+        return getMessageManagerForTrashMailbox(username)
+            .flatMap(messageManager ->
+                deleteExpiredMessages(storeMailboxManager.getSessionProvider().createSystemSession(username),
```

Comment:
Can we move this into a `mailboxSession` variable?

Reply:
sorry I do not get your idea

Reply:
```
MailboxSession mailboxSession = storeMailboxManager.getSessionProvider().createSystemSession(username);
return getMessageManagerForTrashMailbox(username)
    .flatMap(messageManager ->
                deleteExpiredMessages(maiboxSession, 
[...]
```
something like that :)

Reply:
Or even better, you could create that mailboxSession variable at the beginning of your service run, and then pass it down your functions instead of passing username, as you can access the username from mailbox session :)

Reply:
Nah forget that last comment :)